### PR TITLE
Fix some mp3/mp4 media files show as Raw

### PR DESCRIPTION
**Pull request by marxangels**
from

image

to

image

### CODE REVIEW

**Comment by Gusted** [comment]
Should mp3/mp4 be viewed as normal file? IIRC Gitlab/Github still haves them under `/raw`

**Comment by marxangels** [comment]
> Should mp3/mp4 be viewed as normal file? IIRC Gitlab/Github still haves them under `/raw`

It depends on how you use it. 
Maybe we can add a configuration for the end user to make a decision.

**Comment by lunny** [comment]
You can use `display=` like what `svg` file did.

File: modules/setting/mime_type_map.go [hunk 1]
```diff
@@ -21,6 +24,7 @@ func newMimeTypeMap() {
 	m := make(map[string]string, len(keys))
 	for _, key := range keys {
 		m[strings.ToLower(key.Name())] = key.Value()
+		_ = mime.AddExtensionType(key.Name(), key.Value())
```
Comment (Gusted) [inline modules/setting/mime_type_map.go]: I don't think we should ignore a error here.

File: modules/typesniffer/typesniffer.go [hunk 2]
```diff
@@ -87,6 +89,26 @@ func DetectContentType(data []byte) SniffedType {
 	return SniffedType{ct}
 }
 
+func DetectContentTypeExtFirst(name string, readSeekerOrBytes interface{}) (SniffedType, error) {
+	// we can merge `setting.MimeTypeMap` by mime.AddExtensionType()
+	ct := mime.TypeByExtension(filepath.Ext(name))
+	if ct != "" && !strings.Contains(ct, "text/") {
```
Comment (Gusted) [inline modules/typesniffer/typesniffer.go]: Is it possible to change Contains -> HasPrefix?

File: modules/typesniffer/typesniffer.go [hunk 3]
```diff
@@ -87,6 +89,26 @@ func DetectContentType(data []byte) SniffedType {
...
+		return SniffedType{ct}, nil
+	}
+	if r, ok := readSeekerOrBytes.(io.ReadSeeker); ok {
+		st, err := DetectContentTypeFromReader(r)
+		if nil != err {
+			return SniffedType{}, err
+		}
+		_, err = r.Seek(0, io.SeekStart)
```
Comment (Gusted) [inline modules/typesniffer/typesniffer.go]: Why is `Seek` needed?

File: routers/common/repo.go [hunk 4]
```diff
@@ -63,6 +63,7 @@ func ServeData(ctx *context.Context, name string, size int64, reader io.Reader)
 	// Google Chrome dislike commas in filenames, so let's change it to a space
 	name = strings.ReplaceAll(name, ",", " ")
 
+	// st, err := typesniffer.DetectContentTypeExtFirst(name, buf)
```
Comment (silverwind) [inline routers/common/repo.go]: Commented code is confusing and we shouldn't have it. If you want to retain it for future reference, it's better to add a sentence or two about it.

**Comment by marxangels** [comment]
Build is failing. I'll fix this after another pull request #18448 is finished.

File: routers/common/repo.go [hunk 5]
```diff
@@ -21,7 +22,55 @@ import (
 	"code.gitea.io/gitea/modules/util"
 )
 
-// ServeBlob download a git.Blob
+func setCommonHeaders(ctx *context.Context, name string, data interface{}) error {
+	// Google Chrome dislike commas in filenames, so let's change it to a space
+	name = strings.ReplaceAll(name, ",", " ")
```
Comment (Gusted) [inline routers/common/repo.go:27]: I don't think this is needed actually, from what I remember about this weird bug, if you quote the filename(as being done in this function) chromium won't error about it.

**Comment by lunny** [comment]
Please resolve the conflicts.

**Pull request closed without merging**